lib/Model/CMakeLists.txt

```diff
@@ -11,3 +11,17 @@ revng_add_analyses_library_internal(revngModel
 
 target_link_libraries(revngModel
   revngSupport)
+
+set(OUTPUT_DIR "${CMAKE_BINARY_DIR}/include/revng/Model")
+set(MAIN_OUTPUT "${GENERATED_DIR}/Generated.h")
+set(MODEL_YAML "${CMAKE_SOURCE_DIR}/include/revng/Model/schema.yml")
+
+add_custom_command(
+    OUTPUT "${MAIN_OUTPUT}"
+    COMMAND "${CMAKE_SOURCE_DIR}/scripts/model_gen_ng/main.py"
+    ARGS "${MODEL_YAML}" "${OUTPUT_DIR}"
+)
```

Comment:
OK, let's see. I've to say that, while all of this is super cool, when looking at the hand-written files, I'm like "wait, what fields this class has again?".

Maybe it's better to disaggregate the YAML file in the various header files.

```c++
#pragma once

#include "revng/Model/GeneratedSegment.h"

class model::Segment : public model::generated::Segment {
//+ - name: Segment
//+   type: struct
//+   fields:
//+     - name: StartAddress
//+       type: MetaAddress
//+     - name: EndAddress
//+       type: MetaAddress
//+     - name: StartOffset
//+       type: uint64_t
//+     - name: EndOffset
//+       type: uint64_t
//+     - name: IsReadable
//+       type: bool
//+       default: "false"
//+     - name: IsWriteable
//+       type: bool
//+       default: "false"
//+     - name: IsExecutable
//+       type: bool
//+       default: "false"
//+     - name: CustomName
//+       type: Identifier
//+       optional: true
//+
//+   key:
//+     - StartAddress
//+     - EndAddress
public:
  using generated::Segment::Segment;

public:
  Identifier name() const;

public:
  bool verify() const debug_function;
  bool verify(bool Assert) const debug_function;
  bool verify(VerifyHelper &VH) const;
  void dump() const debug_function;
};

#include "revng/Model/GeneratedSegment_late.h"

static_assert(validateTupleTree<model::Segment>(IsYamlizable));
```

Then, in CMake, we explicitly list the header files, we have a script that basically `grep //+` and assembles the YAML file in the build directory, and then the script we already have uses it.

It's not ideal, but overall I think it's better.

Reply:
Having it inlined with the C++ is cool, I only have one thing to note, I would not require prefixing every line with `//+`, it's tedious. We could delimit the YAML with something like `/*<marker> <multiline yaml> <marker>*/`, it can be extracted with a regex just as easily.

Reply:
OK, let's give that a shot. `// TupleTree: Model.yml` and `// End TupleTree`? Or maybe no terminator and just keep eating comments until you find a blank line?